# Next steps {#sec-next}

```{r, echo = FALSE, message = FALSE}
library("tidyverse")
```

**Objectives**

- Introduce the notion of data containers 

- Give an overview of the `SummarizedExperiment`, extensively used in omics analyses


Data in bioinformatics is often more complex than the basic data types
we have seen so far. To deal with this, developers define specialised
data containers (termed classes) that match the properties of the
data they need to handle.

This aspect is central to the **Bioconductor**[^Bioconductor] project which 
uses the same **core data infrastructure** across packages. This certainly contributed 
to Bioconductor's success. Bioconductor package developers are advised to make 
use of existing infrastructure to provide coherence, interoperability and stability to
the project as a whole.

[^Bioconductor]:The [Bioconductor](http://www.bioconductor.org) was initiated by
Robert Gentleman, one of the two creators of the R language. Bioconductor provides 
tools for the analysis and comprehension of omics data. 
Bioconductor uses the R statistical programming language, and is open source and 
open development. 


To illustrate such an omics data container, we'll present the `SummarizedExperiment`
class.

## SummarizedExperiment

The figure below represents the anatomy of SummarizedExperiment.

```{r SE,  echo=FALSE, out.width = '80%'}
knitr::include_graphics("./figs/SE.png")
```


Objects of the class SummarizedExperiment contain :

- **One (or more) assay(s)** containing the quantitative omics data
  (expression data), stored as a matrix-like object. Features (genes,
  transcripts, proteins, ...) are defined along the rows and samples
  along the columns.

- A **sample metadata** slot containing sample co-variates, stored as a
  data frame. Rows from this table represent samples (rows match exactly the 
  columns of the expression data). Columns from this table represent sample 
  covariates along the columns.

- A **feature metadata** slot containing feature co-variates, stored as
  data frame. The rows of this dataframe's match exactly the rows of the 
  expression data. 
  


The coordinated nature of the high throughput data guarantees that the
dimensions of the different slots will always match (i.e the columns
in the expression data and then rows in the sample metadata, as well
as the rows in the expression data and feature metadata) during data
manipulation. 

The metadata slots can grow additional co-variates
(columns) without affecting the other structures.



### Creating a SummarizedExperiment

Let's return to the `rna` dataset that we have used previously.

```{r , message = FALSE}
rna <- read_csv("data/rnaseq.csv")
rna
names(rna)
```

In this table, we can find gene expression per gene and per sample, as well as 
columns describing the samples (orgnanism, age, sex...), and columns describing 
the genes (ENTREZID, product, chromosome_name, gene_biotype...)

We will convert this table in a `SummarizedExperiment`. 
To do so we will extract from this table:

- A count matrix that will be used as the **`assay`**

- A table describing the samples that will be used as the **sample metadata** slot

- A table describing the genes that will be used as the **features metadata** slot

**count matrix**

```{r}
counts <- rna %>%
  select(gene, sample, expression) %>% 
  pivot_wider(names_from = sample,
              values_from = expression) 

count_matrix <- counts %>% select(-gene) %>% as.matrix()
rownames(count_matrix) <- counts$gene

count_matrix[1:5, ]
```

**table describing the samples**

```{r}
sample_metadata <- rna %>%
  select(sample, organism, age, sex, infection, strain, time, tissue, mouse) 

# Many lines are repeated, as in rna, each gene (n = 1474) was analysed in each sample!
sample_metadata

# Remove redundancy
sample_metadata <- unique(sample_metadata)
sample_metadata
```

**table describing the genes**

```{r}
gene_metadata <- rna %>%
  select(gene, ENTREZID, product, ensembl_gene_id, external_synonym, chromosome_name, gene_biotype, phenotype_description, hsapiens_homolog_associated_gene_name)      

# Remove redundancy
gene_metadata <- unique(gene_metadata)
gene_metadata
```


We can now put the different parts together using the
`SummarizedExperiment` constructor:

```{r, message = FALSE}
#BiocManager::install("SummarizedExperiment")
library("SummarizedExperiment")
```

```{r}
se <- SummarizedExperiment(assays = count_matrix,
                           colData = sample_metadata,
                           rowData = gene_metadata)
se
```

Using this data structure, we can access the expression matrix with
the `assay` function, the feature metadata with the `rowData` function,
and the sample metadata with the `colData` function:


```{r}
dim(assay(se))
head(assay(se))
```

```{r}
colData(se)
```

```{r}
dim(rowData(se))
head(rowData(se))
```

#### Adding variables to metadata

We can also add information to the metadata.
Suppose that you want to add the center where the samples were collected...

```{r}
colData(se)$center <- rep("University of Illinois", nrow(colData(se)))
colData(se)
```

This illustrates that the metadata slots can grow indefinitely without affecting 
the other structures!


### Subsetting a SummarizedExperiment

SummarizedExperiment can be subset just like with data frames, 
with numerics or with characters of logicals.

Below, we create a new instance of class SummarizedExperiment that contains only 
the 5 first features for the 3 first samples.

```{r}
se1 <- se[1:5, 1:3]
se1
```

We can also use the colData() function to subset on something from the sample metadata. 
For example, here we keep only the non infected samples.

```{r}
se1 <- se[, colData(se)$infection == "NonInfected"]
se1
colData(se1)
```

Of course the rowData() function can similarly be used to keep only certain 
features of interest. We could for example focus on miRNAs... 

```{r}
se1 <- se[rowData(se)$gene_biotype == "miRNA",]
se1
assay(se1)
```


`r msmbstyle::question_begin()`

Extract the gene expression levels of the 3 first genes in sample at time 0 and at time 8.

`r msmbstyle::question_end()`

`r msmbstyle::solution_begin()`

```{r, purl=FALSE}
assay(se)[1:3, colData(se)$time != 4]

# Equivalent to
assay(se)[1:3, colData(se)$time == 0 | colData(se)$time == 8]
```
`r msmbstyle::solution_end()`

**Take-home message**


- `SummarizedExperiment` represent an efficient way to store and to handle omics 
data.

- They are used in many Bioconductor packages. 

If you follow next training focused on RNA sequencing analysis, you will learn to 
use the Bioconductor `DESeq2` package to do some differential expression analyses. 
`DESeq2`'s whole analysis is handled in a `SummarizedExperiment`.


